### PR TITLE
fix: Transform scoping for `for X` in loop

### LINKED ISSUE #16382: [Bug]: Transform block scoping fails with for in statement inside loop
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

I was able to replicate this issue in babeljs.io with the following code:

```js
function w() {
    for (let i = 0; i < y; i++) {
        let outer = {}
        for (var key in someObj) {
            let x = () => outer;
        }
    }
}
```

### Configuration file name

_No response_

### Configuration

Using the transform-block-scoping plugin only.

### Current and expected behavior

Expected Babel to be able to parse & transform the code without throwing any exceptions.

Actual behavior is an error:
```
/repl.js: Property left of ForInStatement expected node to be of a type ["VariableDeclaration","LVal"] but instead got undefined
```

### Environment

I could replicate this issue in babeljs.io

### Possible solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by liuxingbaoyu**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #16382 
| Patch: Bug Fix?          | √
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | √
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

### CODE REVIEW

File: packages/babel-plugin-transform-block-scoping/test/fixtures/general/for-for-x/input.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+function w() {
+    for (let i = 0; i < y; i++) {
+        let outer = {}
+        for (var key in someObj) {
+            let x = () => outer;
+        }
+    }
+}
+
+function w2() {
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-block-scoping/test/fixtures/general/for-for-x/input.js:10]: What is the difference between these two functions?

File: packages/babel-plugin-transform-block-scoping/src/loop.ts [hunk 2]
```diff
@@ -239,17 +239,13 @@ export function wrapLoopBody(
...
+        assign.push(decl.id as t.Identifier);
       }
     }
     if (assign.length > 0) {
-      let replacement: t.Node =
+      const replacement: t.Node =
         assign.length === 1 ? assign[0] : t.sequenceExpression(assign);
-      if (
-        !t.isForStatement(varPath.parent, { init: varPath.node }) &&
-        !t.isForXStatement(varPath.parent, { left: varPath.node })
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-block-scoping/src/loop.ts:249]: Can you also add a test swapping the `for ... in` and `for (;;)` loops?

File: packages/babel-plugin-transform-block-scoping/test/fixtures/general/for-for-x/input.js (hunk 1 above)
Comment (liuxingbaoyu) [reply packages/babel-plugin-transform-block-scoping/test/fixtures/general/for-for-x/input.js:10]: One is `for in` and the other is `for of`. :)

Comment (nicolo-ribaudo) [reply packages/babel-plugin-transform-block-scoping/test/fixtures/general/for-for-x/input.js:10]: Ohh I played the "find the difference between the two pictures" for like 3 minutes before giging up :grimacing:

**Review by nicolo-ribaudo — APPROVED** [review]

**Pull request merged**